File: c++/95_Unique_Binary_Search_Trees_II.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    int numTrees(int n) {
    	int dp[n + 1] = {0};
    	dp[0] = 1;
    	for (int i = 1; i <= n; i++)
    		for (int j = 0; j < i; j++)
    			dp[i] += dp[j] * dp[i - 1 - j];
    	return dp[n];
    }
};



int main() {
	Solution s;
}

File: python/95_Unique_Binary_Search_Trees_II.py
#!/usr/bin/python
import unittest

class Solution(object):
    def generateTrees(self, n):
        if n == 0:
            return []
        return self.dfs(1, n)
    
    def dfs(self, Tmin, Tmax):
    	if Tmin > Tmax:
    		return [None]
    	resList = []
    	# generate leftSubTree and rightSubTree
    	for k in range(Tmin, Tmax + 1):
    		leftSubTree = self.dfs(Tmin, k - 1)
    		rightSubTree = self.dfs(k + 1, Tmax)

	    	for i in leftSubTree:
	    		for j in rightSubTree:
	    			rootNote = TreeNode(k)
	    			rootNote.left = i
	    			rootNote.right = j
	    			resList.append(rootNote)
    	return resList
        


class TestSolution(unittest.TestCase):
    def setUp(self):
        self.s = Solution()

    def test_method(self):
        """Such as self.assertEqual, self.assertTrue"""
        self.assertEqual(self.s.numTrees(1), 1)
        self.assertEqual(self.s.numTrees(2), 2)
        pass


if __name__ == "__main__":
    unittest.main()